### PR TITLE
Add support for Psalm v5

### PR DESCRIPTION
**Pull request by mcaskill**
This pull request continues what was started in #41 and should be merged first.

---

Currently, the dependency constraints target vimeo/psalm `^5@beta` and psalm/plugin-phpunit `dev-master`.

Fixed errors and warnings raised by Psalm.

@kkmuffme has [mentioned](https://github.com/humanmade/psalm-plugin-wordpress/pull/41#issuecomment-1293095342) they also have a Psalm v5 ready fork of this plugin, which I was not aware of when I started my version. I'm sharing my implementation for reference and can be replaced by theirs.

### CODE REVIEW

File: .editorconfig [hunk 1]
```diff
@@ -0,0 +1,18 @@
+root = true
```
Comment (ntwb) [inline .editorconfig:1]: See my comment in #41 about this, though this will get updated once #41 is merged of course

**Review by ntwb — COMMENTED** [review]
A quick glance and this looks good, lets get #41 merged and this PR updated and go from them there

**Comment by mcaskill** [comment]
@kkmuffme I would like your input on this. Even see your implementation.

File: composer.json [hunk 2]
```diff
@@ -15,12 +15,12 @@
 		"php-stubs/wordpress-stubs": "^5.5",
 		"php-stubs/wordpress-globals": "^0.2.0",
 		"php-stubs/wp-cli-stubs": "^2.5",
-		"vimeo/psalm": "^4"
+		"vimeo/psalm": "^4 || ^5 || ^5@beta"
```
Comment (ntwb) [inline composer.json:18]: Unsure of the feasability of this, but maybe we should try and configure GitHub Actions to test each version

Currently, `vimeo/psalm (5.0.0-beta1)` is being installed

https://github.com/humanmade/psalm-plugin-wordpress/actions/runs/3342040266/jobs/5533851486#step:3:79

**Comment by mcaskill** [comment]
The new workflow appears to work well but there's a breaking change introduced in Psalm 5 that will requiring either a breaking change for the plugin or some kind of conditional statement.

> [`TypeError: Psalm\Storage\FunctionLikeParameter::__construct(): Argument #6 ($type_location) must be of type ?Psalm\CodeLocation, bool given`](https://github.com/humanmade/psalm-plugin-wordpress/actions/runs/3342206656/jobs/5534196072#step:6:21)

The `FunctionLikeParameter::__construct()` method has received a new parameter since the release of 5.0.0-beta1:

References:

* [Psalm 4.x](https://github.com/vimeo/psalm/blob/4.x/src/Psalm/Storage/FunctionLikeParameter.php#L116-L126)
* [Psalm 5.0.0-beta1](https://github.com/vimeo/psalm/blob/5.0.0-beta1/src/Psalm/Storage/FunctionLikeParameter.php#L116-L126)
* [Psalm 5.x](https://github.com/vimeo/psalm/blob/c2ce7e6/src/Psalm/Storage/FunctionLikeParameter.php#L116-L128)

To preserve compatibility, we could use the `PSALM_VERSION` constant but it might not be available and will vary a lot depending on the tag or branch of Psalm that is installed.
Alternatively, we can use `ReflectionMethod` to lookup the constant and detect the index of the `bool $is_optional = true` parameter that the plugin sets to false.

**Comment by ntwb** [comment]
> Anyway regarding v5:
> do we really need to preserve v4 compatibility? Even psalm itself will drop it.
> I think the better way would be to release one last working (!) v4 version of this plugin.

I agree, what I think we need to ensure is just that, a v4 that we know as works in the v2.x.x branch, and then:

> And then the next release (with the release of psalm v5) will be v5 only. I don't think investing scarce contributor time in unnecessary compatibility stuff is worth it.

This new v5 Psalm release would be a _major_ bump to a new release v3.x.x version

Users would need to explicitly opt-in to v3.x.x

**Comment by mcaskill** [comment]
@kkmuffme Hello, when you have an opportunity: could you create a branch to share your fork's changes? Alternatively, privately share me a copy or add me as a collaborator to your fork, and I can extract out what's needed for Psalm v5.

Thanks.

**Comment by ZebulanStanphill** [comment]
@kkmuffme Psalm 5.15.0 released 2 weeks ago.

**Comment by paulshryock** [comment]
Is there any list of outstanding work needed to complete this?

**Comment by paulshryock** [comment]
> Really sorry, didn't have time bc of family the last months.
> 
> It's ready actually - atm there is a pending PR to be merged by me in core https://github.com/vimeo/psalm/pull/10370 which fixes tons of false negative/false positive errors that are especially relevant for WP/this plugin.
> 
> I will PR the new version of this plugin this week and release it beginning of next week with the next psalm core version. 
> In the meantime there have been tons of improvements and features added to this plugin, I just didn't find time to clean it up and release it.
> 

Awesome, thank you! And I certainly don't mean to rush you. I know it's the holiday season, and totally understandable to have a lot going on with family. Just seeing if I could help out in any way. Sounds like you have everything just about ready. 👍❤️

**Comment by mcaskill** [comment]
Closing in favour of #49.

**Pull request closed without merging**